core/fetcher.go

```diff
@@ -54,41 +73,86 @@ func (f *BlockFetcher) GetBlockInfo(ctx context.Context, height *int64) (*types.
 }
 
 // GetBlock queries Core for a `Block` at the given height.
+// if the height is nil, use the latest height
 func (f *BlockFetcher) GetBlock(ctx context.Context, height *int64) (*types.Block, error) {
-	res, err := f.client.Block(ctx, height)
+	blockHeight, err := f.resolveHeight(ctx, height)
 	if err != nil {
 		return nil, err
 	}
 
-	if res != nil && res.Block == nil {
-		return nil, fmt.Errorf("core/fetcher: block not found, height: %d", height)
+	stream, err := f.client.BlockByHeight(ctx, &coregrpc.BlockByHeightRequest{Height: blockHeight})
+	if err != nil {
+		return nil, err
 	}
-
-	return res.Block, nil
+	block, _, _, _, err := receiveBlockByHeight(stream)
+	if err != nil {
+		return nil, err
+	}
+	return block, nil
 }
 
 func (f *BlockFetcher) GetBlockByHash(ctx context.Context, hash libhead.Hash) (*types.Block, error) {
-	res, err := f.client.BlockByHash(ctx, hash)
+	if hash == nil {
+		return nil, fmt.Errorf("cannot get block with nil hash")
+	}
+	stream, err := f.client.BlockByHash(ctx, &coregrpc.BlockByHashRequest{Hash: hash})
 	if err != nil {
 		return nil, err
 	}
-
-	if res != nil && res.Block == nil {
-		return nil, fmt.Errorf("core/fetcher: block not found, hash: %s", hash.String())
+	block, err := receiveBlockByHash(stream)
+	if err != nil {
+		return nil, err
 	}
 
-	return res.Block, nil
+	return block, nil
+}
+
+// resolveHeight takes a height pointer and returns its value if it's not nil.
+// otherwise, returns the latest height.
+func (f *BlockFetcher) resolveHeight(ctx context.Context, height *int64) (int64, error) {
+	if height != nil {
+		return *height, nil
+	}
+	status, err := f.client.Status(ctx, &coregrpc.StatusRequest{})
```

Comment:
Isn't there a way to request the latest height through existing methods? Let's avoid additional roundtrip for Head request if possible.

Reply:
Like `0` on `BlockByHeightRequest` can represent latest height.

Reply:
That's why I recommended to just use 0

Reply:
Im fine with changing and removing the pointers. Is this something you want to do in this PR?

Reply:
Yes pls, we would like to leave as minimal tech debt behind as possiblr

Reply:
7a6257044bf23f4c88e6a243ba8bb31a5da6def4

Reply:
So just to confirm, you've changed this on the tendermint gRPC server-side such that requesting height 0 is the same as requesting height == nil over the RPC client ( return latest block )?

Reply:
https://github.com/celestiaorg/celestia-core/pull/1513/commits/6291264239f8c3cb3f5add5da7a9e6742422ab85